### ISSUE TITLE
Dice coefficient

### ISSUE THREAD

**Issue by ramchandracheke** [frankkramer-lab/MIScnn#79]
Hi,

Thanks for your amazing code. Does it support calculation of dice score without background?

Thanks,
Ram

**Comment by muellerdo** [comment]
PS: Wait do you need it for model training or for evaluation purposes?

If you need it for evaluation then I would recommend to check out this work of us for 3D covid-19 segmentation.
In this evaluation script, I loaded the ground truth and prediction via MIScnn, and then compute the DSC for each class.
You can simply exclude the background results then.

https://github.com/frankkramer-lab/covid19.MIScnn/blob/master/scripts/run_evaluation.py

**Comment by ramchandracheke** [comment]
Hi @muellerdo 

Thank you for your guidance. I am training a model using model.evaluate function. Yes, one of my friend used MISCNN framework for his masters project. I found it very helpful rather than coding from scratch. I am trying to recreate a scenario where the background is ignored and avg dice score for 2 classes were calculated per epoch. Since background is always start with high dice score, this is little bit misleading in terms of comparing result with their implementation. 

Thanks for your suggestion on Focal loss, that's interesting! Please let me know once that implement goes live. I am happy to try it.

https://docs.monai.io/en/latest/losses.html#diceloss They have included (include_background=True,) term in dice loss. If I am able to code it for keras then will create a pull request.

Thanks,
Ram

**Comment by ramchandracheke** [comment]
Hi,

I have written a function for dice coefficient with No background. It assume network has BHWD[N]
Batch, Height, Width, Depth, No of channels.

def dice_coefficient_nb(y_true, y_pred, smooth=0.00001):
    y_true=y_true[:,:,:,:,1:]
    y_pred=y_pred[:,:,:,:,1:]
    
    y_true_f = K.flatten(y_true)
    y_pred_f = K.flatten(y_pred)
    intersection = K.sum(y_true_f * y_pred_f)
    return (2. * intersection + smooth) / \
           (K.sum(y_true_f) + K.sum(y_pred_f) + smooth)

def dice_coefficient_loss_nb(y_true, y_pred):
    return 1-dice_coefficient_nb(y_true, y_pred)
